controller/src/main/java/io/pravega/controller/server/eventProcessor/requesthandlers/CommitRequestHandler.java

```diff
@@ -387,4 +407,15 @@ public CommitRequestHandler(final StreamMetadataStore streamMetadataStore,
         return streamMetadataStore.getState(event.getScope(), event.getStream(), true, null, executor)
                                   .thenApply(state -> state.equals(State.COMMITTING_TXN) || state.equals(State.SEALING));
     }
+
+    /* This is a data class that represents a writer mark. Writers send mark information
+     * for watermarking purposes, containing time and position.
+     */
+    @Data
+    @AllArgsConstructor
+    public static class TxnWriterMark {
```

Comment:
FYI this class is not thread safe. You should mark it as such (`@NotThreadSafe`) or make it thread safe if you pass it between threads or work concurrently on it. This is even more important since you marked it as `public` which means you intend on giving it to outside components.

Reply:
this should be a data class, immutable.. all the fields in this class should be final.

Reply:
fixed. Made the class immuatble and moved to AbstractStreamMetadataStore.

Reply:
> FYI this class is not thread safe. You should mark it as such (`@NotThreadSafe`) or make it thread safe if you pass it between threads or work concurrently on it. This is even more important since you marked it as `public` which means you intend on giving it to outside components.

I needed to mark it as public as it is used outside its parent package by CommitRequestHandler.